mayastor/src/bdev/dev/nvmx/controller.rs

```diff
@@ -0,0 +1,1335 @@
+                    Running,
+                    Faulted(ControllerFailureReason::ResetFailed),
+                );
+            }
+
+            // Unlock the controller before calling the callback to avoid
+            // potential deadlocks.
+            drop(controller);
+        }
+
+        (reset_ctx.cb)(status == 0, reset_ctx.cb_arg);
+    }
+
+    fn _reset_destroy_channels(
+        channel: &mut NvmeIoChannelInner,
+        ctx: &mut ResetCtx,
+    ) -> i32 {
+        debug!("Resetting I/O channel");
+
+        // Bail out preliminary if shutdown is active.
+        if ctx.shutdown_in_progress {
+            return 0;
+        }
+
+        // Check in advance for concurrent controller shutdown.
+        if channel.is_shutdown() {
+            ctx.shutdown_in_progress = true;
+            return 0;
+        }
+
+        let rc = channel.reset();
+
+        if rc == 0 {
+            debug!("I/O channel successfully reset");
+        } else {
+            error!("failed to reset I/O channel (rc={}), reset aborted", rc);
+        }
+        rc
+    }
+
+    fn _reset_destroy_channels_done(status: i32, reset_ctx: ResetCtx) {
+        if status != 0 {
+            error!(
+                "{}: controller reset failed with status = {}",
+                reset_ctx.name, status
+            );
+            NvmeController::_complete_reset(reset_ctx, status);
+            return;
+        }
+
+        info!("{}: all I/O channels successfully reset", reset_ctx.name);
+        // In case shutdown is active, don't reset the controller as its
+        // being removed.
+        if reset_ctx.shutdown_in_progress {
+            info!(
+                "{}: controller shutdown detected, skipping reset",
+                reset_ctx.name
+            );
+            return;
+        }
+
+        let rc = unsafe { spdk_nvme_ctrlr_reset(reset_ctx.spdk_handle) };
+        if rc != 0 {
+            error!(
+                "{} failed to reset controller, rc = {}",
+                reset_ctx.name, rc
+            );
+
+            NvmeController::_complete_reset(reset_ctx, rc);
+        } else {
+            info!(
+                "{} controller successfully reset, reinitializing I/O channels",
+                reset_ctx.name
+            );
+
+            /* Once controller is successfully reset, schedule another I/O
+             * channel traversal to restore all I/O channels.
+             */
+            let io_device = Arc::clone(&reset_ctx.io_device);
+            io_device.traverse_io_channels(
+                NvmeController::_reset_create_channels,
+                NvmeController::_reset_create_channels_done,
+                reset_ctx,
+            );
+        }
+    }
+
+    fn _reset_create_channels(
+        channel: &mut NvmeIoChannelInner,
+        reset_ctx: &mut ResetCtx,
+    ) -> i32 {
+        // Make sure no cuncurrent shutdown takes place.
+        if channel.is_shutdown() {
+            return 0;
+        }
```

Comment:
Is it possible for a shutdown to occur after this check?

Reply:
It should not no. Once shut down, channels should not be recreated because we want to shut down. So the logic behind, here should not be run.